Bottom: d4bfce8f4c6be977c8d490433a4625cc572856e7
Top:    29b9e08fe9d44d2f701a118b8342fe31b545f0f1
Author: Ian Kent <raven@themaw.net>
Date:   2013-08-19 08:27:40 +0800

autofs-5.0.7 - fix bad mkdir permission on create

Reported by Gordon Lack (gordoni[dot]m[dot]lackr[at]gsk[dot]com).

If the automount daemon needs to create a directory (hierarchy) for an
automount and it is started up with a umask of 027 (or similar) then it
creates unusable directories (permission == 550).


---

diff --git a/daemon/automount.c b/daemon/automount.c
index 97c726a..6b55103 100644
--- a/daemon/automount.c
+++ b/daemon/automount.c
@@ -122,7 +122,10 @@ static int do_mkdir(const char *parent, const char *path, mode_t mode)
 		status = statfs(parent, &fs);
 	if ((status != -1 && fs.f_type == (__SWORD_TYPE) AUTOFS_SUPER_MAGIC) ||
 	    contained_in_local_fs(path)) {
-		if (mkdir(path, mode) == -1) {
+		int mask = umask(0022);
+		int ret = mkdir(path, mode);
+		void umask(mask);
+		if (ret == -1) {
 			errno = EACCES;
 			return 0;
 		}
